Generation: customization for Storage

niander opened the pull request.

markcowl commented on src/Storage/custom/hybrid-2019-03-01/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,338 @@
+            $null = $PSBoundParameters.Remove("KeyVaultEncryption")
+        }
+        $null = Az.Storage.internal\New-AzStorageAccount @PSBoundParameters
+        if ($PSBoundParameters.ContainsKey("InputObject")) {
```
Per above, there should be no InputObject for New cmdlets

markcowl commented on src/Storage/custom/hybrid-2019-03-01/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,338 @@
+            $PSBoundParameters["InputObject"] | Az.Storage.internal\Get-AzStorageAccountProperty
+        } else {
+            Az.Storage.internal\Get-AzStorageAccountProperty -Name $PSBoundParameters["Name"] -ResourceGroupName $PSBoundParameters["ResourceGroupName"]
+        }
```
Is this replacing the command, or are there additional aprameter sets?

Right now it looks like there should at least be an additional parameter set where neither Storage nor KeyVault encryption is used.

niander replied:
If no switch is used (`-StorageEncryption` or `-KeyVaultEncryption`) it would have no encryption. I found another parameter set to be unnecessary

markcowl commented on src/Storage/custom/hybrid-2019-03-01/Update-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,373 @@
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Category('Body')]
+        [System.String]
+        # Gets or sets the custom domain name assigned to the storage account. Name is the CNAME source.
+        ${CustomDomainName},
```
This parameter cannot be mandatory, and it cna only be in parameter sets that include encryption

niander replied:
Ditto

markcowl commented on src/Storage/custom/hybrid-2019-03-01/Update-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,373 @@
+        # Allows https traffic only to storage service if sets to true.
+        ${EnableHttpsTrafficOnly},
+    
+        #[Parameter(ParameterSetName='UpdateExpanded', Mandatory, HelpMessage='The encryption keySource (provider). Possible values (case-insensitive): Microsoft.Storage, Microsoft.Keyvault')]
```
remove commented code

markcowl commented on src/Storage/custom/hybrid-2019-03-01/Update-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,373 @@
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Category('Body')]
+        [System.Management.Automation.SwitchParameter]
+        # A boolean indicating whether or not the service encrypts the data as it is stored.
+        ${FileEnabled},
```
`EncryptFileService`  This should only be in parameter sets that include encryption, right?

niander replied:
In the original cmdlets these "Enabled" encryptions are never assigned. I will try to test those.

markcowl commented on src/Storage/custom/hybrid-2019-03-01/Update-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,373 @@
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Category('Body')]
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Support.DefaultAction]
+        # Specifies the default action of allow or deny when no other rules match.
+        ${NetworkRuleSetDefaultAction},
```
This cannot be mandatory

markcowl commented on src/Storage/custom/hybrid-2019-03-01/Update-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,373 @@
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Category('Body')]
+        [System.Management.Automation.SwitchParameter]
+        # A boolean indicating whether or not the service encrypts the data as it is stored.
+        ${QueueEnabled},
```
`EncryptQueueService`  This should only be in parameter sets where encryption is enabled

markcowl commented on src/Storage/custom/hybrid-2019-03-01/Update-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,373 @@
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Category('Body')]
+        [System.Management.Automation.SwitchParameter]
+        # A boolean indicating whether or not the service encrypts the data as it is stored.
+        ${TableEnabled},
```
`EncryptTableService` again, only in encryption parameter sets

markcowl commented on src/Storage/custom/hybrid-2019-03-01/Update-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,373 @@
+            $null = $PSBoundParameters.Add("EncryptionKeySource", "Microsoft.Keyvault")
+            $null = $PSBoundParameters.Remove("KeyVaultEncryption")
+        }
+        if ($PSBoundParameters.ContainsKey("NoEncryption")) {
```
A few changes are needed here:

(1) Need to pass the `DontShow` parameters into the Get cmdlet
(2) Cannot simply pass through the AsJob parameter.  In this case, we need to use Start-Job amd execute the rest of the parameters (without AsJob) synchronously inside the job

markcowl commented on src/Storage/custom/latest-2019-04-30/Get-AzStorageAccount_GetByName.ps1:
```diff
@@ -0,0 +1,75 @@
+    [OutputType('Microsoft.Azure.PowerShell.Cmdlets.Storage.Models.Api20190401.IStorageAccount')]
+    [CmdletBinding(PositionalBinding=$false)]
+    [Microsoft.Azure.PowerShell.Cmdlets.Storage.Profile('latest-2019-04-30')]
+    [Microsoft.Azure.PowerShell.Cmdlets.Storage.Description('Lists all the storage accounts available under the subscription. Note that storage keys are not returned; use the ListKeys operation for this.')]
```
Gets a single storage account

markcowl commented on src/Storage/custom/latest-2019-04-30/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,356 @@
+    [OutputType('Microsoft.Azure.PowerShell.Cmdlets.Storage.Models.Api20190401.IStorageAccount')]
+    [CmdletBinding(DefaultParameterSetName='CreateExpandedStorageEncryption', PositionalBinding=$false, SupportsShouldProcess, ConfirmImpact='Medium')]
+    [Microsoft.Azure.PowerShell.Cmdlets.Storage.Profile('latest-2019-04-30')]
+    [Microsoft.Azure.PowerShell.Cmdlets.Storage.Description('Asynchronously creates a new storage account with the specified parameters. If an account is already created and a subsequent create request is issued with different properties, the account properties will be updated. If an account is already created and a subsequent create or update request is issued with the exact same set of properti ...
```
does not need `Asynchronously`

markcowl commented on src/Storage/custom/latest-2019-04-30/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,356 @@
+    [Microsoft.Azure.PowerShell.Cmdlets.Storage.Profile('latest-2019-04-30')]
+    [Microsoft.Azure.PowerShell.Cmdlets.Storage.Description('Asynchronously creates a new storage account with the specified parameters. If an account is already created and a subsequent create request is issued with different properties, the account properties will be updated. If an account is already created and a subsequent create or update request is issued with the exact same set of properti ...
+    param(
+        [Parameter(ParameterSetName='CreateExpandedStorageEncryption', Mandatory, HelpMessage='The name of the storage account within the specified resource group. Storage account names must be between 3 and 24 characters in length and use numbers and lower-case letters only.')]
```
See comments from `hybrid` parofile cmdlet - they still apply

markcowl commented on src/Storage/custom/latest-2019-04-30/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,356 @@
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Category('Path')]
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Models.IStorageIdentity]
+        # Identity Parameter
+        ${InputObject},
```
As above, no InputObject here

markcowl commented on src/Storage/custom/latest-2019-04-30/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,356 @@
+        [Parameter(ParameterSetName='CreateExpandedStorageEncryption', HelpMessage='Specify IdentityType as ''SystemAssigned''')]
+        [Parameter(ParameterSetName='CreateExpandedKeyVaultEncryption', HelpMessage='Specify IdentityType as ''SystemAssigned''')]
+        [Parameter(ParameterSetName='CreateViaIdentityExpandedStorageEncryption', HelpMessage='Specify IdentityType as ''SystemAssigned''')]
+        [Parameter(ParameterSetName='CreateViaIdentityExpandedKeyVaultEncryption', HelpMessage='Specify IdentityType as ''SystemAssigned''')]
```
As above, no ViaIdentity parameter sets

markcowl commented on src/Storage/custom/latest-2019-04-30/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,356 @@
+.Link
+https://docs.microsoft.com/en-us/powershell/module/az.storage/new-azstorageaccount
+#>
+function New-AzStorageAccount {
```
As above we seem to need a parameter set with no encryption

niander replied:
As I said I don't think we need because they are switch parameters. What do you think?

markcowl commented on src/Storage/custom/latest-2019-04-30/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,356 @@
+        # Allows https traffic only to storage service if sets to true.
+        ${EnableHttpsTrafficOnly},
+    
+        #[Parameter(ParameterSetName='CreateExpanded', Mandatory, HelpMessage='The encryption keySource (provider). Possible values (case-insensitive): Microsoft.Storage, Microsoft.Keyvault')]
```
remove commented code

markcowl commented on src/Storage/custom/latest-2019-04-30/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,356 @@
+        [Parameter(ParameterSetName='CreateViaIdentityExpandedKeyVaultEncryption', Mandatory, HelpMessage='Required. Gets or sets the location of the resource. This will be one of the supported and registered Azure Geo Regions (e.g. West US, East US, Southeast Asia, etc.). The geo region of a resource cannot be changed once it is created, but if an identical geo region is specified on update, the ...
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Category('Body')]
+        [System.String]
+        # Required. Gets or sets the location of the resource. This will be one of the supported and registered Azure Geo Regions (e.g. West US, East US, Southeast Asia, etc.). The geo region of a resource cannot be changed once it is created, but if an identical geo region is specified on update, the request will succeed.
```
needs a location completer, we will add these later

markcowl commented on src/Storage/custom/latest-2019-04-30/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,356 @@
+            $null = $PSBoundParameters.Add("EncryptionKeySource", "Microsoft.Keyvault")
+            $null = $PSBoundParameters.Remove("KeyVaultEncryption")
+        }
+        $null = Az.Storage.internal\New-AzStorageAccount @PSBoundParameters
```
see comments on Update method above - we need to pass int he `DontSHow` parameters to the Get below, and we need to handle AsJob for both

markcowl commented on src/Storage/custom/latest-2019-04-30/Set-AzRmStorageContainerImmutabilityPolicy.ps1:
```diff
@@ -0,0 +1,122 @@
+.Link
+https://docs.microsoft.com/en-us/powershell/module/az.storage/set-azrmstoragecontainerimmutabilitypolicy
+#>
+function Set-AzRmStorageContainerImmutabilityPolicy {
```
This should have a parameter set that accepts the storage account object

niander replied:
As discussed we agreed on not having a parameter set for just storage accounts here.

markcowl commented on src/Storage/custom/latest-2019-04-30/Update-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,384 @@
+https://docs.microsoft.com/en-us/powershell/module/az.storage/update-azstorageaccount
+#>
+function Update-AzStorageAccount {
+    [OutputType('Microsoft.Azure.PowerShell.Cmdlets.Storage.Models.Api20190401.IStorageAccount')]
```
Apply comments as above

markcowl commented on src/Storage/custom/latest-2019-04-30/Update-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,384 @@
+        # Allows https traffic only to storage service if sets to true.
+        ${EnableHttpsTrafficOnly},
+    
+        #[Parameter(ParameterSetName='UpdateExpanded1', Mandatory, HelpMessage='The encryption keySource (provider). Possible values (case-insensitive): Microsoft.Storage, Microsoft.Keyvault')]
```
remove commented code

markcowl commented on src/Storage/custom/latest-2019-04-30/Update-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,384 @@
+        if ($PSBoundParameters.ContainsKey("NoEncryption")) {
+            $null = $PSBoundParameters.Add("EncryptionKeySource", "")
+            $null = $PSBoundParameters.Remove("NoEncryption")
+        }
```
Same comments as above on AsJob and common parameters

markcowl commented on src/Storage/custom/hybrid-2019-03-01/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,338 @@
+        if ($PSBoundParameters.ContainsKey("KeyVaultEncryption")) {
+            $null = $PSBoundParameters.Add("EncryptionKeySource", "Microsoft.Keyvault")
+            $null = $PSBoundParameters.Remove("KeyVaultEncryption")
+        }
```
See notes below on update - for cmdlets that use AsJob and string together multiple cmdlets, we will need to follow a slightly different pattern

niander commented on src/Storage/custom/hybrid-2019-03-01/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,338 @@
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Category('Body')]
+        [System.Management.Automation.SwitchParameter]
+        # A boolean indicating whether or not the service encrypts the data as it is stored.
+        ${BlobEnabled},
```
Done.

niander commented on src/Storage/custom/hybrid-2019-03-01/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,338 @@
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Category('Body')]
+        [System.String]
+        # Gets or sets the custom domain name assigned to the storage account. Name is the CNAME source.
+        ${CustomDomainName},
```
It should be because the Name property is mandatory but CustomDomain object is not. I filled an issue: https://github.com/Azure/autorest.powershell/issues/408

markcowl replied:
But, since this is a customization, we can make it optional, can't we?

niander commented on src/Storage/custom/hybrid-2019-03-01/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,338 @@
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Category('Body')]
+        [System.Management.Automation.SwitchParameter]
+        # A boolean indicating whether or not the service encrypts the data as it is stored.
+        ${FileEnabled},
```
Done.

markcowl commented on src/Storage/custom/hybrid-2019-03-01/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,275 @@
+        [Parameter(ParameterSetName='CreateExpanded1KeyVaultEncryption', HelpMessage='Specifies whether traffic is bypassed for Logging/Metrics/AzureServices. Possible values are any combination of Logging|Metrics|AzureServices (For example, "Logging, Metrics"), or None to bypass none of those traffics.')]
+        [ArgumentCompleter([Microsoft.Azure.PowerShell.Cmdlets.Storage.Support.Bypass])]
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Category('Body')]
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Support.Bypass]
```
Will need a type converter for this to support AsJob

markcowl commented on src/Storage/custom/hybrid-2019-03-01/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,275 @@
+        [Parameter(ParameterSetName='CreateExpanded1KeyVaultEncryption', HelpMessage='Specifies the default action of allow or deny when no other rules match.')]
+        [ArgumentCompleter([Microsoft.Azure.PowerShell.Cmdlets.Storage.Support.DefaultAction])]
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Category('Body')]
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Support.DefaultAction]
```
Same

markcowl commented on src/Storage/custom/hybrid-2019-03-01/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,275 @@
+    process {
+        if ($PSBoundParameters.ContainsKey("AsJob")) {
+            $null = $PSBoundParameters.Remove("AsJob")
+            Start-Job -ScriptBlock {param($params) Az.Storage\New-AzStorageAccount @params} -ArgumentList $PSBoundParameters
```
This still needs the Import-Module call as in the PR I sent

markcowl commented on src/Storage/custom/latest-2019-04-30/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,292 @@
+        [Parameter(ParameterSetName='CreateExpandedKeyVaultEncryption', HelpMessage='Specifies whether traffic is bypassed for Logging/Metrics/AzureServices. Possible values are any combination of Logging|Metrics|AzureServices (For example, "Logging, Metrics"), or None to bypass none of those traffics.')]
+        [ArgumentCompleter([Microsoft.Azure.PowerShell.Cmdlets.Storage.Support.Bypass])]
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Category('Body')]
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Support.Bypass]
```
As above, needs a type converter

markcowl commented on src/Storage/custom/latest-2019-04-30/New-AzStorageAccount.ps1:
```diff
@@ -0,0 +1,292 @@
+        [Parameter(ParameterSetName='CreateExpandedStorageEncryption', HelpMessage='Specifies the default action of allow or deny when no other rules match.')]
+        [Parameter(ParameterSetName='CreateExpandedKeyVaultEncryption', HelpMessage='Specifies the default action of allow or deny when no other rules match.')]
+        [ArgumentCompleter([Microsoft.Azure.PowerShell.Cmdlets.Storage.Support.DefaultAction])]
+        [Microsoft.Azure.PowerShell.Cmdlets.Storage.Category('Body')]
```
same

niander commented:
@markcowl All these AsJob support changes that were pending I had already implemented last week but I forgot to update my fork. The changes were saved just in the working directory. I just updated with it all. Please take a look
